src/contracts/yas_factory.cairo

```diff
@@ -0,0 +1,251 @@
+                FeeAmountEnabled { fee: 10000, tick_spacing: IntegerTrait::<i32>::new(200, false) }
+            );
+    }
+
+    #[external(v0)]
+    impl YASFactoryImpl of IYASFactory<ContractState> {
+        fn owner(self: @ContractState) -> ContractAddress {
+            self.owner.read()
+        }
+
+        fn fee_amount_tick_spacing(self: @ContractState, fee: u32) -> i32 {
+            self.fee_amount_tick_spacing.read(fee)
+        }
+
+        fn pool(
+            self: @ContractState, token_a: ContractAddress, token_b: ContractAddress, fee: u32
+        ) -> ContractAddress {
+            self.pool.read((token_a, token_b, fee))
+        }
+
+        fn create_pool(
+            ref self: ContractState, token_a: ContractAddress, token_b: ContractAddress, fee: u32
+        ) -> ContractAddress {
+            assert(token_a != token_b, 'tokens must be different');
+            assert(
+                token_a.is_non_zero() && token_b.is_non_zero(), 'tokens addresses cannot be zero'
+            );
+
+            let (token_0, token_1) = if token_a < token_b {
+                (token_a, token_b)
+            } else {
+                (token_b, token_a)
+            };
+
+            let tick_spacing = self.fee_amount_tick_spacing(fee);
+            assert(tick_spacing.is_non_zero(), 'tick spacing not initialized');
+
+            assert(self.pool(token_0, token_1, fee).is_zero(), 'token pair already created');
+
+            let contract_address_salt = self.generate_salt(@token_0, @token_1);
+            let calldata = self.serialize_calldata(@token_0, @token_1, fee, @tick_spacing);
+
+            let (pool, _) = deploy_syscall(
+                self.pool_class_hash.read(), contract_address_salt, calldata.span(), false
+            )
+                .unwrap_syscall();
+
+            self.pool.write((token_0, token_1, fee), pool);
+            // populate mapping in the reverse direction, deliberate choice to avoid the cost of comparing addresses
+            self.pool.write((token_1, token_0, fee), pool);
+
+            self.emit(PoolCreated { token_0, token_1, fee, tick_spacing, pool });
+
+            pool
+        }
+
+        fn enable_fee_amount(ref self: ContractState, fee: u32, tick_spacing: i32) {
+            self.assert_only_owner();
+            assert(fee < 1000000, 'fee cannot be gt 1000000');
+
+            let zero = IntegerTrait::<i32>::new(0, false);
+            // tick spacing is capped at 16384 to prevent the situation where tick_spacing is so large that
+            // TickBitmap#nextInitializedTickWithinOneWord overflows int24 container from a valid tick
+            // 16384 ticks represents a >5x price change with ticks of 1 bips
+            assert(
+                tick_spacing > zero && tick_spacing < IntegerTrait::<i32>::new(16384, false),
+                'wrong tick_spacing (0<ts<16384)'
+            );
+            assert(self.fee_amount_tick_spacing(fee) == zero, 'fee amount already initialized');
```

Comment:
```suggestion
            // tick spacing is capped at 16384 to prevent the situation where tick_spacing is so large that
            // TickBitmap#nextInitializedTickWithinOneWord overflows int24 container from a valid tick
            // 16384 ticks represents a >5x price change with ticks of 1 bips
            assert(
                tick_spacing > Zeroable::zero() && tick_spacing < IntegerTrait::<i32>::new(16384, false),
                'wrong tick_spacing (0<ts<16384)'
            );
            assert(self.fee_amount_tick_spacing(fee).is_zero(), 'fee amount already initialized');
```